File: Core/Model/Impl/LocCanChangeDirectionPredicate.cs
namespace BinkyRailways.Core.Model.Impl
{
    /// <summary>
    /// Predicate that evaluates to true the given loc is allowed to change direction.
    /// </summary>
    public sealed class LocCanChangeDirectionPredicate : LocPredicate, ILocCanChangeDirectionPredicate
    {
        /// <summary>
        /// Accept a visit by the given visitor
        /// </summary>
        public override TReturn Accept<TReturn, TData>(EntityVisitor<TReturn, TData> visitor, TData data)
        {
            return visitor.Visit(this, data);
        }

        /// <summary>
        /// Create a deep clone.
        /// </summary>
        protected internal override LocPredicate Clone()
        {
            return new LocCanChangeDirectionPredicate();
        }

        /// <summary>
        /// Human readable name of this type of entity.
        /// </summary>
        public override string TypeName
        {
            get { return Strings.LocChangeChangeDirectionPredicateTypeName; }
        }

        /// <summary>
        /// Convert to a string
        /// </summary>
        public override string ToString()
        {
            return Strings.LocCanChangeDirectionPredicateDescription;
        }
    }
}

File: pkg/core/model/impl/loc_can_change_direction_predicate.go
// Copyright 2020 Ewout Prangsma
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author Ewout Prangsma
//

package impl

import "github.com/binkyrailways/BinkyRailways/pkg/core/model"

// LocCanChangeDirectionPredicate extends implementation methods to model.LocCanChangeDirectionPredicate
type LocCanChangeDirectionPredicate interface {
	ModuleEntity
	model.LocCanChangeDirectionPredicate
}

type locCanChangeDirectionPredicate struct {
	locPredicate
}

var _ model.LocCanChangeDirectionPredicate = &locCanChangeDirectionPredicate{}

// newLocCanChangeDirectionPredicate creates a new can-change-direction predicate
func newLocCanChangeDirectionPredicate() *locCanChangeDirectionPredicate {
	p := &locCanChangeDirectionPredicate{}
	return p
}

// Accept a visit by the given visitor
func (p *locCanChangeDirectionPredicate) Accept(v model.EntityVisitor) interface{} {
	return v.VisitLocCanChangeDirectionPredicate(p)
}

// Create a deep clone.
func (p *locCanChangeDirectionPredicate) Clone() model.LocPredicate {
	return newLocCanChangeDirectionPredicate()
}

// Evaluate this predicate for the given loc.
func (p *locCanChangeDirectionPredicate) Evaluate(loc model.Loc) bool {
	return loc.GetChangeDirection() == model.ChangeDirectionAllow
}
